ext/socket/lib/socket.rb

```diff
@@ -623,46 +711,126 @@ def accept_nonblock(exception: true)
   def self.tcp(host, port, local_host = nil, local_port = nil, connect_timeout: nil, resolv_timeout: nil) # :yield: socket
     last_error = nil
     ret = nil
+    attempt_v6 = true
+    attempt_v4 = true
+    start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+
+    if connect_timeout
+      raise ArgumentError, "connect_timeout must be Numeric" unless connect_timeout.is_a?(Numeric)
+      raise ArgumentError, "connect_timeout must not be negative" if connect_timeout.negative?
+    end
 
     local_addr_list = nil
     if local_host != nil || local_port != nil
       local_addr_list = Addrinfo.getaddrinfo(local_host, local_port, nil, :STREAM, nil)
+      attempt_v6 = local_addr_list.any? {|local_ai| local_ai.afamily == Socket::AF_INET6 }
+      attempt_v4 = local_addr_list.any? {|local_ai| local_ai.afamily == Socket::AF_INET }
+    end
+
+    queue = Queue.new
+
+    if attempt_v6
+      getaddrinfo_v6_th = start_getaddrinfo_v6(host, port, queue, resolv_timeout)
     end
 
-    Addrinfo.foreach(host, port, nil, :STREAM, timeout: resolv_timeout) {|ai|
-      if local_addr_list
-        local_addr = local_addr_list.find {|local_ai| local_ai.afamily == ai.afamily }
-        next unless local_addr
+    if attempt_v4
+      getaddrinfo_v4_th = start_getaddrinfo_v4(host, port, queue, resolv_timeout)
+    end
+
+    if connect_timeout
+      timeout_th = Thread.new do
+        sleep(connect_timeout)
+        queue.close()
+      end
+    end
+
+    socket_list = []
+    getaddrinfo_v6_done = attempt_v6 ? false : true
+    getaddrinfo_v4_done = attempt_v4 ? false : true
+    while ai = queue.pop do
```

Comment:
`queue.pop` must not block except the first.

Consider the following scenario:

- IPv6 getaddrinfo takes 1 second
- IPv6 connect takes 1 second
- IPv4 getaddrinfo takes 10 second

In this scenario, the first `queue.pop` needs 1 second to wait IPv6 getaddrinfo.  This is not a problem.
IPv6 connect_nonblock returns :wait_writable.
IO.select on the IPv6 socket timeouts CONNECTION_ATTEMPT_DELAY = 0.25 second
(IPv6 connect will finish 1 - 0.25 = 0.75 second after the timeout)
Then, the second `quque.pop` needs 10 + RESOLUTION_DELAY - (1 + 0.25) = 8.8 second
But IPv6 connect finishes at 0.75 second after the timeout of CONNECTION_ATTEMPT_DELAY

So, this implementation wastes 8.8 - 0.75 = 8.05 second.
This is not acceptable.

This method must multiplex notification from getaddrinfo threads AND writability of connecting sockets.
It is difficult because IO.select cannot wait Queue.

One idea is using a pipe instead of Queue.
Another idea is to consult with ko1.

Reply:
I fixed it to use IO.pipe instead of Queue.